src/events/SDL_pen.c

```diff
@@ -0,0 +1,365 @@
+{
+    PEN_LOAD(pen, penid, NULL);
+    return pen->name;
+}
+
+Uint32
+SDL_PenCapabilities(SDL_PenID penid)
+{
+    PEN_LOAD(pen, penid, 0u);
+    return pen->flags & ~(SDL_PEN_FLAG_STALE | SDL_PEN_FLAG_INACTIVE);
+}
+
+Uint32
+SDL_PenStatus(SDL_PenID penid, float * x, float * y, float * axes, size_t num_axes)
+{
+    PEN_LOAD(pen, penid, 0u);
+
+    if (x) {
+        *x = pen->last_x;
+    }
+    if (y) {
+        *y = pen->last_y;
+    }
+    if (axes && num_axes) {
+        size_t axes_to_copy = SDL_min(num_axes, SDL_PEN_NUM_AXES);
+        SDL_memcpy(axes, pen->last_axes, sizeof(float) * axes_to_copy);
+    }
+    return pen->last_status;
+}
+
+void
+SDL_PenStringForGUID(SDL_PenGUID guid, char *pszGUID, int cbGUID)
+{
+    SDL_JoystickGetGUIDString(guid, pszGUID, cbGUID);
+}
+
+SDL_PenGUID
+SDL_PenGUIDForString(const char *pchGUID)
+{
+    return SDL_JoystickGetGUIDFromString(pchGUID);
+}
+
+void
+SDL_PenGCMark(void)
+{
+    int i;
+    for (i = 0; i < pen_handler.pens_used; ++i) {
+        SDL_Pen *pen = &pen_handler.pens[i];
+        pen->flags |= SDL_PEN_FLAG_STALE;
+    }
+    pen_handler.unsorted = SDL_TRUE;
+}
+
+SDL_Pen *
+SDL_PenRegister(SDL_PenID id, SDL_PenGUID guid, char *name, Uint32 capabilities)
+{
+    const size_t alloc_growth_constant = 1;  /* Expect few pens */
+
+    SDL_Pen *pen = SDL_GetPen(id.id);
+    if (!pen) {
+        if (!pen_handler.pens || pen_handler.pens_used == pen_handler.pens_allocated) {
+            size_t pens_to_allocate = pen_handler.pens_allocated + alloc_growth_constant;
+            SDL_Pen *pens;
+            if (pen_handler.pens) {
+                pens = SDL_realloc(pen_handler.pens, sizeof(SDL_Pen) * pens_to_allocate);
+                SDL_memset(pens + pen_handler.pens_used, 0,
+                           sizeof(SDL_Pen) * (pens_to_allocate - pen_handler.pens_allocated));
+            } else {
+                pens = SDL_calloc(sizeof(SDL_Pen), pens_to_allocate);
+            }
+            pen_handler.pens = pens;
+            pen_handler.pens_allocated = pens_to_allocate;
+        }
+        pen = &pen_handler.pens[pen_handler.pens_used];
+        pen_handler.pens_used += 1;
+    }
+    pen->id = id;
+    pen->flags = capabilities & (~SDL_PEN_FLAG_STALE);
+    pen->guid = guid;
+    SDL_strlcpy(pen->name, name, SDL_PEN_MAX_NAME);
+
+    return pen;
+}
+
+void
+SDL_PenGCSweep(void (*free_deviceinfo)(void*))
+{
+    int i;
+    /* We don't actually free the SDL_Pen entries, so that we can still answer queries about
+       formerly active SDL_PenIDs later.  */
+    for (i = 0; i < pen_handler.pens_used; ++i) {
+        SDL_Pen *pen = &pen_handler.pens[i];
+        if (pen->flags & SDL_PEN_FLAG_STALE) {
+            pen->flags |= SDL_PEN_FLAG_INACTIVE;
+            if (pen->deviceinfo) {
+                free_deviceinfo(pen->deviceinfo);
+                pen->deviceinfo = NULL;
+            }
+        }
+        pen->flags &= ~SDL_PEN_FLAG_STALE;
+    }
+    qsort(pen_handler.pens,
```

Comment:
`SDL_qsort`

Don't know what to do with `bsearch`, though.

Reply:
> Don't know what to do with `bsearch`, though.

Maybe this implementation can be added as `SDL_bsearch` to SDL's stdlib:
https://github.com/DevSolar/pdclib/blob/master/functions/stdlib/bsearch.c

Reply:
That `bsearch` looks fine to me, thank you for both pieces of feedback!
I'll temporarily add a custom (slightly less clever one).
I can submit a PR for `bsearch` but first need to figure out how test automation works in SDL...

Reply:
I have a patch for bsearch.

@slouken, @icculus: Are we interested in a `SDL_bsearch`? (for 2.0.22 or not.)